prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs

```diff
@@ -1846,20 +1842,117 @@ layoutWadlerLeijen
+               apples
+               Donaudampfschifffahrtskapitänskajütenmülleimer
+
+Before #83 was fixed, that wasn't what we got! Instead we got this:
+
+> renderIO stdout $ layoutSmart opts doc
+Groceries: pommes
+           de
+           terre
+           apples
+           Donaudampfschifffahrtskapitänskajütenmülleimer
+
+Why?
+
+minNestingLevel was effectively defined as
+
+    minNestingLevel = lineIndent
+
+The lineIndent for "pommes de terre" is 0.
+
+The FittingPredicate for layoutSmart will continue to check the rest of the
+document until it finds a line where the indentation <= minNestingLevel.
+In this case this meant that layoutSmart would traverse all the items,
+and note that the last item, Donaudampfschifffahrtskapitänskajütenmülleimer,
+doesn't fit into the available space! The "flatter" version of the document
+has failed, so "pommes de terre" gets spread over several lines!
+
+Obviously this would be an inconsistency with the layout of the other items.
+Their lineIndent is 11 each, so for them, the FittingPredicate stops already
+on the next line.
+
+The obvious solution is to change the definition of minNestingLevel:
+
+    minNestingLevel = currentColumn
+
+This however breaks the "python-ish" document from the documentation for
+layoutSmart:
+
+    expected: |------------------------|
+              fun(
+                fun(
+                  fun(
+                    fun(
+                      fun(
+                        [ abcdef
+                        , ghijklm ])))))
+              |------------------------|
+
+     but got: |------------------------|
+              fun(
+                fun(
+                  fun(
+                    fun(
+                      fun([ abcdef
+                          , ghijklm ])))))
+              |------------------------|
+
+We now accept the worse layout because the problematic last line has
+the same indentation as the current column of "[ abcdef", so we don't check it!
+
+The solution we went with in the end is a bit of a hack:
+
+We check whether the alternative, "high" layout is a (potentially less wide)
+hanging layout, and in that case pick its indentation as the minNestingLevel.
+
+This way we achieve the optimal layout in both scenarios.
+
+See https://github.com/quchen/prettyprinter/issues/83 for the bug that lead
+to the current solution.
+
+
+Note [Detecting failure with Unbounded page width]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+To understand why it is sufficient to check the first line of the
+SimpleDocStream, trace how an SFail ends up there:
+
+1. We group a Doc containing a Line, producing a (Union x y) where
+   x contains Fail.
+
+2. In best, any Unions are handled recursively, rejecting any
+   alternatives that would result in SFail.
+
+So once a SimpleDocStream reaches selectNicer, any SFail in it must
+appear before the first linebreak – any other SFail would have been
+detected and rejected in a previous iteration.
+-}
```

Comment:
Sounds good. A bit long, but probably apt for explaining this relatively special case.